legend-sdlc-server-fs/src/main/java/org/finos/legend/sdlc/server/api/entity/FileSystemEntityApi.java

```diff
@@ -0,0 +1,1118 @@
+            String committerName = commit.getCommitterIdent().getName();
+            Instant committedTimeStamp = commit.getCommitterIdent().getWhenAsInstant();
+            String message = commit.getFullMessage();
+            return new FileSystemRevision(revisionID, authorName, authoredTimeStamp, committerName, committedTimeStamp, message);
+        }
+    }
+
+    public static class FileSystemProjectFileFileModificationContext implements ProjectFileAccessProvider.FileModificationContext
+    {
+        private final String projectId;
+        private final String revisionId;
+        private final SourceSpecification sourceSpecification;
+
+        public FileSystemProjectFileFileModificationContext(String projectId, SourceSpecification sourceSpecification, String revisionId)
+        {
+            this.projectId = projectId;
+            this.revisionId = revisionId;
+            this.sourceSpecification = Objects.requireNonNull(sourceSpecification, "source specification may not be null");
+        }
+
+        @Override
+        public Revision submit(String message, List<? extends ProjectFileOperation> operations)
+        {
+            String branchName = getRefBranchName(sourceSpecification);
+            try
+            {
+                int changeCount = operations.size();
+                List<LocalCommitAction> commitActions = operations.stream().map(this::fileOperationToCommitAction).collect(Collectors.toCollection(() -> Lists.mutable.ofInitialCapacity(changeCount)));
+                String referenceRevisionId = this.revisionId;
+                if (referenceRevisionId != null)
+                {
+                    String targetBranchRevision = FileSystemRevision.getFileSystemRevision(this.projectId, branchName).getId();
+                    if (!referenceRevisionId.equals(targetBranchRevision))
+                    {
+                        String msg = "Expected " + sourceSpecification + " to be at revision " + referenceRevisionId + "; instead it was at revision " + targetBranchRevision;
+                        LOGGER.info(msg);
+                        throw new LegendSDLCServerException("Exp", Response.Status.CONFLICT);
+                    }
+                }
+                Repository repo = FileSystemWorkspaceApi.retrieveRepo(this.projectId);
+                Git git = new Git(repo);
+                git.checkout().setName(branchName).call();
+                for (LocalCommitAction commitAction : commitActions)
+                {
+                    switch (commitAction.getAction())
+                    {
+                        case CREATE:
+                        {
+                            File newFile = new File(repo.getDirectory().getParent(), commitAction.getFilePath());
+                            newFile.getParentFile().mkdirs();
+                            newFile.createNewFile();
+                            Files.write(newFile.toPath(), commitAction.getContent());
```

Comment:
This will be updated accordingly, but no need to handle the exception for the case where file already exists as we check for this when we build ProjectFileOperations. We check if the file exists, if so we do throw an exception.